### PR TITLE
feat(preferences): Allow to define preferences through environment variables

### PR DESCRIPTION
**Pull request by benoitf**
### What does this PR do?
Allow to define extra preferences from environment variables in sidecars (that could come from meta.yaml files)

env variable name is `CHE_THEIA_SIDECAR_PREFERENCES`
env variable value is `stringified value of a JSON content`

### Screenshot/screencast of this PR

https://user-images.githubusercontent.com/436777/112611244-de515500-8e1d-11eb-9dd5-0814636bc3ee.mp4

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/17975
https://github.com/eclipse/che/issues/19199

### How to test this PR?
Add CHE_THEIA_SIDECAR_PREFERENCES in a meta.yaml env section and look if the preferences is added

here is a devfile example:
```yaml
apiVersion: 1.0.0
metadata:
  name: custom-env
attributes:
  multiRoot: 'on'
components:
  - memoryLimit: 512Mi
    type: cheEditor
    reference: 'https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-1047/che_theia_meta.yaml'
    alias: che-theia
  - type: chePlugin
    reference: 'https://gist.githubusercontent.com/benoitf/ba757696b6591f792487b7a006f21d15/raw/a789b4013053ddbcbd938c3b77a85dd98277d7b9/asciidoctor.yaml'

```

Look at preferences you should see that you've a preference named `asciidoc.helloworld:false`

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=next

Change-Id: I2761540bf307d00319b61e77e3203fe803742c28
Signed-off-by: Florent Benoit <fbenoit@redhat.com>

### CODE REVIEW

**Comment by benoitf** [comment]
I've updated description with test example

**Review by azatsarynnyy — APPROVED** [review]
Thank you for the test Devfile!
I've just tested it - works as expected
image

File: extensions/eclipse-che-theia-preferences-provider-extension/src/browser/prefs-provider.ts [hunk 1]
```diff
@@ -32,19 +32,37 @@ export class PreferencesProvider implements FrontendApplicationContribution {
...
+      if (component.plugin) {
+        return component.plugin.preferences || {};
+      }
+      return {};
+    });
+
+    const componentStatusPreferences = componentStatuses.map(componentStatus => {
+      // check if env var is matching
+      if (componentStatus?.env) {
+        const preferencesEnv = componentStatus.env.find(env => env.name === 'CHE_THEIA_SIDECAR_PREFERENCES');
```
Comment (sunix) [inline extensions/eclipse-che-theia-preferences-provider-extension/src/browser/prefs-provider.ts:51]: So the env variable is not used as an env variable.

Comment (benoitf) [reply extensions/eclipse-che-theia-preferences-provider-extension/src/browser/prefs-provider.ts:51]: it's a marker yes for che-server usecase.

For DevWorkspace usecase, preferences will be read directly from che-theia-plugins.yaml files

**Pull request merged**